RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT.cc

```diff
@@ -0,0 +1,778 @@
+}
+
+bool DAClusterizerInZT::merge(vector<vertex_t> & y, int nt)const{
+  // merge clusters that collapsed or never separated, return true if vertices were merged, false otherwise
+  
+  if(y.size()<2)  return false;
+  
+  for(vector<vertex_t>::iterator k=y.begin(); (k+1)!=y.end(); k++){
+    if( std::abs( (k+1)->z - k->z ) < 1.e-3 &&
+        std::abs( (k+1)->t - k->t ) < 1.e-3    ){  // with fabs if only called after freeze-out (splitAll() at highter T)
+      double rho = k->pk + (k+1)->pk;
+      if(rho>0){
+        k->z = ( k->pk * k->z + (k+1)->z * (k+1)->pk)/rho;
+        k->t = ( k->pk * k->t + (k+1)->t * (k+1)->pk)/rho;
+      }else{
+        k->z = 0.5*(k->z + (k+1)->z);
+        k->t = 0.5*(k->t + (k+1)->t);
+      }
+      k->pk = rho;
+      
+      y.erase(k+1);
+      return true;  
+    }
+  }
+  
+  return false;
+}
+
+bool DAClusterizerInZT::merge(vector<vertex_t> & y, double & beta)const{
+  // merge clusters that collapsed or never separated, 
+  // only merge if the estimated critical temperature of the merged vertex is below the current temperature
+  // return true if vertices were merged, false otherwise
+  if(y.size()<2)  return false;
+
+  for(vector<vertex_t>::iterator k=y.begin(); (k+1)!=y.end(); k++){
+    if ( std::abs((k+1)->z - k->z) < 2.e-3 && 
+         std::abs((k+1)->t - k->t) < 2.e-3    ) { 
+      double rho=k->pk + (k+1)->pk;
+      double swE=k->swE+(k+1)->swE - k->pk * (k+1)->pk / rho * ( std::pow((k+1)->z - k->z,2.) + 
+                                                                 std::pow((k+1)->t - k->t,2.)   );
+      double Tc=2*swE/(k->sw+(k+1)->sw);
+      
+      if(Tc*beta<1){
+	if(rho>0){
+	  k->z = ( k->pk * k->z + (k+1)->z * (k+1)->pk)/rho;
+          k->t = ( k->pk * k->t + (k+1)->t * (k+1)->pk)/rho;
+	}else{
+	  k->z = 0.5*(k->z + (k+1)->z);
+          k->t = 0.5*(k->t + (k+1)->t);
+	}
+	k->pk  = rho;
+	k->sw += (k+1)->sw;
+	k->swE = swE;
+	k->Tc  = Tc;
+	y.erase(k+1);
+	return true; 
+      }
+    }
+  }
+
+  return false;
+}
+
+bool DAClusterizerInZT::purge(vector<vertex_t> & y, vector<track_t> & tks, double & rho0, const double beta)const{
+  // eliminate clusters with only one significant/unique track
+  if(y.size()<2)  return false;
+  
+  unsigned int nt=tks.size();
+  double sumpmin=nt;
+  vector<vertex_t>::iterator k0=y.end();
+  for(vector<vertex_t>::iterator k=y.begin(); k!=y.end(); k++){ 
+    int nUnique=0;
+    double sump=0;
+    double pmax=k->pk/(k->pk+rho0*exp(-beta*dzCutOff_*dzCutOff_));
+    for(unsigned int i=0; i<nt; i++){
+      if(tks[i].Z > 0){
+	double p = k->pk * std::exp(-beta*Eik(tks[i],*k)) / tks[i].Z ;
+	sump+=p;
+	if( (p > 0.9*pmax) && (tks[i].pi>0) ){ nUnique++; }
+      }
+    }
+
+    if((nUnique<2)&&(sump<sumpmin)){
+      sumpmin=sump;
+      k0=k;
+    }
+  }
+ 
+  if(k0!=y.end()){
+    if(verbose_){cout << "eliminating prototype at " << k0->z << "," << k0->t 
+                      << " with sump=" << sumpmin << endl;}
+    //rho0+=k0->pk;
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant